Added a typecast to int before doing the parameter check for spin locks

Lyle-Alloy opened the pull request:
This prevents a comparison between a signed and an unsigned number which
will create a warning tripping -Werror.

This is my fix to #306

kilograham commented on src/rp2_common/hardware_sync/include/hardware/sync.h:
```diff
@@ -205,7 +205,7 @@ inline static spin_lock_t *spin_lock_instance(uint lock_num) {
 inline static uint spin_lock_get_num(spin_lock_t *lock) {
     int lock_num = lock - (spin_lock_t *) (SIO_BASE + SIO_SPINLOCK0_OFFSET);
-    invalid_params_if(SYNC, lock_num < 0 || lock_num >= NUM_SPIN_LOCKS);
+    invalid_params_if(SYNC, lock_num < 0 || lock_num >= (int) NUM_SPIN_LOCKS);
```
would make more sense to make `lock_num` a `uint` from the start an skip the < 0 check.

Lyle-Alloy replied:
I think I actually like it how it is right now, because I think the intent was to check if lock was either pointing below the `spin_lock` block or pointing above it, which is reflected in the two conditionals.

Making `lock_num` a `uint` would achieve the same thing, but it would be because if it was below the `spin_lock` block `lock_num` would roll over on subtraction, which is kinda opaque to the casual reader in my opinion.

Anyways, you're the maintainer and thanks for the work! Don't want to add more mental load than needed so I'll just make the change.

lurch replied:
I agree that we ought to try and avoid under/over flows! Perhaps it's worth adding a separate check for `lock >= (spin_lock_t *) (SIO_BASE + SIO_SPINLOCK0_OFFSET)` ? :man_shrugging:

Lyle-Alloy commented:
@kilograham @lurch How's this? I just moved all the checks into the `invalid_params_if` and made them very human readable.

All those constants should be evaluated at compile time and merged together so it shouldn't be much work at runtime.

I also added a check for alignment because that would be undefined behaviour as well!

kilograham commented:
cool thanks, yeah i noticed that our `-Werror` checks aren't run with `PARAM_ASSERTIONS_ENABLE_ALL=1` so I will fix the others

kilograham commented:
can you please rebase onto develop

Lyle-Alloy commented:
@kilograham should be all set

kilograham commented:
thank you... sorry to mess you around, but do you mind rebasing onto develop-1.1.2 now; i want to squeeze it in there actually!

Lyle-Alloy commented:
Haha you're making me remember my `git rebase`-foo. Should be all set now
